tests/experiment.js

```diff
@@ -12,19 +12,123 @@
 */
 
 var stageLoader = require('../core/stage-loader');
+var captureController = require('../lib/test-phases').controller;
+var assert = require('chai').assert;
 
 function testPipeline(stageList, incb) {
   var cb = function(data) { incb(); };
-  stageLoader.processStages(stageList, cb, function(e) { throw e; });
+  var err = function(e) { throw e; };
+  stageLoader.startPipeline(stageList).then(incb);
 }
 
+function experiment(name) {
+  return [{name: 'input', options: {data: name + '.erlnmyr'}}, 'fileToString', 'doExperiment'];
+}
+
+function wentThrough(event, name) {
+  for (var i = 0; i < event.tags.trace.length; i++) {
+    if (event.tags.trace[i].phase.name == name)
+      return true;
+  }
+  return false;
+}
+
+function assertPath(event, list) {
+  assert(event.tags.trace.length <= list.length);
+  for (var i = 0; i < list.length; i++)
+    assert(event.tags.trace[i].phase.name == list[i]);
+}
+
+
 describe('experiments', function() {
-  it('should run to completion', function(done) {
-    testPipeline([
-        {name: 'input', options: {data: 'tests/experiment'}},
-        'readDir',
-        'fileToBuffer',
-        'bufferToString',
-        'doExperiment'].map(stageLoader.stageSpecificationToStage), done);
+
+  before(function() {
+    process.chdir('tests/experiment');
+  });
+  after(function() {
+    process.chdir('../..');
+  });
+
+  it('should be able to branch', function(done) {
+    testPipeline(experiment('simple-branch'), function() {
+      var events = captureController.getInvocations();
+      assert(events.length == 2);
+      assert(events[0].name == events[1].name, "data was duplicated in branch");
+      assert(wentThrough(events[0], 't1To1') !== wentThrough(events[1], 't1To1'),
+          "only one copy of data went through each branch");
+      done();
+    });
+  });
+
+  it('should deal with a harder branch scenario (1)', function(done) {
+    testPipeline(experiment('harder-branch-1'), function() {
+      var events = captureController.getInvocations();
+      assert(events.length == 2);
+      assert(events[0].name == events[1].name, "data was duplicated in branch");
+      if (wentThrough(events[0], 't1To1')) {
+        assertPath(events[0], ['t0ToN', 't1To1', 't1To1']);
+        assert(!wentThrough(events[1], 't1To1'))
+      } else {
+        assertPath(events[1], ['t0ToN', 't1To1', 't1To1']);
+        assert(!wentThrough(events[0], 't1To1'))
+      }
+      done();
+    });
+  });
+
+  it('should deal with a harder branch scenario (2)', function(done) {
+    testPipeline(experiment('harder-branch-2'), function() {
+      var events = captureController.getInvocations();
+      assert(events.length == 2);
+      assert(events[0].name == events[1].name, "data was duplicated in branch");
+      if (wentThrough(events[0], 't1To1')) {
```

Comment:
It is, but I don't know which order is *right* yet. The test can stay flexible until we work that out, I think.